### PR TITLE
Update dependencies to fix vulnerability regarding CVE-2024-23342

### PR DESCRIPTION
**Pull request by vagas87**
python-keycloak replaced python-jose with jwcrypto [recently](https://github.com/marcospereirampj/python-keycloak/commit/e285f92e6ba9a33ab3c93b50ad2577974f1c2fb5), because it uses ecdsa, which is deemed to be [unsafe](https://github.com/advisories/GHSA-wj6h-64fc-37mp). As fastapi-keycloak-middleware still catches exceptions from python-jose, I got import errors in one of my company's projects using this package when I tried to update our dependencies to address this issue. Therefore, I decided to fix this problem from my private GitHub account.

Please release an update for fastapi-keycloak-middleware including this PR as soon as possible. Otherwise we will have to substitute the package in our company's project and I really don't want to deal with all that hustle 😅

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -8,9 +8,9 @@ packages = [{ include = "fastapi_keycloak_middleware" }]
 
 [tool.poetry.dependencies]
 python = "^3.10"
-fastapi = ">=0.73.0"
+fastapi = ">=0.110.0"
```
Comment (waza-ari) [inline pyproject.toml]: As this library should be used in other projects which may have other dependencies, we want to specify the dependencies as much as possible. I don't consider this a task of this library to educate people on keeping their FastAPI version up to date, rather I'd like to specify the minimum compatible version. May I therefore ask you to revert this change?

File: pyproject.toml [hunk 2]
```diff
@@ -8,9 +8,9 @@ packages = [{ include = "fastapi_keycloak_middleware" }]
...
 install = "^1.3.5"
-python-keycloak = ">2.14.0,<4"
+python-keycloak = "^3.9.1"
```
Comment (waza-ari) [inline pyproject.toml:13]: We should make sure that we support newer versions within the same major version. I'd prefer `">3.9.1,<4"` instead

**Comment by waza-ari** [comment]
Hi @vagas87, thanks for your contribution! I'm happy to change to jwcrypto, however I have a few comments regarding the version constraints. Please check the review and let me know what you think

File: pyproject.toml (hunk 1 above)
Comment (vagas87) [reply pyproject.toml]: That's okay, because my current pain is caused by python-jose. However, reverting it in the pyproject.toml as is will not prevent poetry from raising the version to latest in the poetry.lock when updating the dependencies. So at least the CVEs are going to be addressed in your build 😊

File: pyproject.toml (hunk 2 above)
Comment (vagas87) [reply pyproject.toml:13]: "^" does exactly the same, because it won't allow changes to the left most non-zero digit in the version tag (see [Poertry documentation](https://python-poetry.org/docs/dependency-specification/#caret-requirements)). If you prefer the explicit declaration instead, I would change it to ">=3.9.1, <4" or ">3.9.0,<4" because 3.9.1 is the newest version 😅

File: pyproject.toml (hunk 1 above)
Comment (vagas87) [reply pyproject.toml]: See [7dbdf22](https://github.com/waza-ari/fastapi-keycloak-middleware/pull/29/commits/7dbdf22ead1ea3b1689ca666c3c46f9eb0d3e69d)

File: pyproject.toml (hunk 2 above)
Comment (waza-ari) [reply pyproject.toml:13]: There's times I should stick to sleeping... you're right of course, that's fine.

**Comment by waza-ari** [comment]
Thanks again for your contribution. I looked into it a bit more, there's some backwards incompatibilities where I'm not sure how to handle them.

**Audience claim**

The default behaviour of `python-jose` was that the `aud` claim was not required, but would be validated if present. Compare the part in [their code base](https://github.com/mpdavis/python-jose/blob/master/jose/jwt.py).

`jwcrypto` doesn't expose that much control anymore, so `python-keycloak` have made an effort to somehow map the [old `verify_aud` option](https://github.com/marcospereirampj/python-keycloak/blob/2125d1e788a5153eb8ec769cee691edf2324fe8d/src/keycloak/keycloak_openid.py#L546). This results in a change in default behaviour:

With `verify_aud` set to `True`, the authentication now fails if no `aud` claim is present, which used to work in the old version as `require_aud` was `False`. In fact that applies to pretty much all the claims. In my case that did break the application, I don't consider this a minor issue

**JWT Decode Options**

So far we've been exposing the JWT decode options to the user, see [also the docs](https://fastapi-keycloak-middleware.readthedocs.io/en/latest/advanced_topics.html#jwt-decoding-options). This wouldn't be possible anymore, as it would not have any effect with the new version of `python-keycloak`. We'd need to rewrite the doc section and somehow hope no one is making use of these options.

Around the same topic there's still [ongoing discussions](https://github.com/marcospereirampj/python-keycloak/issues/503) in the upstream repo around backwards incompatibility. Especially the fact that enabling `verify_aud` (which is currently the default) prohibits `exp` and `nbf` validation does concern me.

To summarize, I do see the need for the change, but I'd rather have the upstream `python-keycloak` changes settled before releasing a new version.

File: pyproject.toml (hunk 1 above)
Comment (waza-ari) [reply pyproject.toml]: Luckily the lock file is not taken into account at all when this library is being used as dependency. We do require FastAPI at the moment, although it's really only used for type hinting. I don't want to restrict the users regarding the FastAPI version they can use the library with. Thanks for the change.

**Comment by vagas87** [comment]
That's unfortunate, but I understand your concerns. Would it at least be possible to release a patch version that temporarily bounds `python-keycloak` to below v3.9.1 in the pyproject.toml until these problems are solved?

Right now, everyone is getting import errors when including `fastapi-keycloak-middleware` and running anything, because of the package incompatibilities that the dependencies don't account for. The vulnerability (if there even is a real threat, I don't know, I'm just hunting down Dependabot warnings right now for an upcoming release in my company 🙈) would be there nonetheless, but at least nobody is scratching heads wondering why their code isn`t running anymore after trying to fix the CVE.

**Comment by waza-ari** [comment]
Yes absolutely, I have that planned for later today, once I’m back at home.

**Comment by vagas87** [comment]
Awesome, that helps a lot. Thanks 👍

**Comment by waza-ari** [comment]
@vagas87 released version 0.3.1 with the pinned version. if you don't mind we can keep the PR open and I'll work on that once the upstream discussions are settled

**Comment by vagas87** [comment]
Sure, why not.

**Comment by waza-ari** [comment]
Thanks again for your contribution. I've taken parts of it and incorporated it into the v1.0.0 release. Its currently being published on test PyPi, I'll probably release it later today after some more testing.

**Comment by vagas87** [comment]
That's great, thanks you for that 🚀
